MiniApp/Classes/MiniAppDownloader.swift

```diff
@@ -22,52 +18,70 @@ class MiniAppDownloader: NSObject, MiniAppDownloaderProtocol {
+        self.miniAppClient = apiClient
+        self.manifestDownloader.fetchManifest(apiClient: apiClient, appId: appId, versionId: versionId) { (result) in
+            switch result {
+            case .success(let responseData):
+                self.completionHandler = completionHandler
+                self.downloadManifestFiles(with: appId, versionId: versionId, manifestResponse: responseData)
+            case .failure(let error):
+                return completionHandler(.failure(error))
             }
+        }
+    }
 
-            urlToDirectoryMap[url] = pathDirectory.appendingPathComponent(fileDirectory)
+    private func downloadManifestFiles(with appId: String, versionId: String, manifestResponse: ManifestResponse) {
+        guard let miniAppStoragePath = FileManager.getMiniAppDirectory(with: appId, and: versionId) else {
+            return
+        }
+        miniAppDirectory = miniAppStoragePath
+        downloadMiniApp(manifestResponse.files, to: miniAppStoragePath)
+    }
 
+    private func downloadMiniApp(_ urls: [String], to miniAppPath: URL) {
+        self.miniAppClient?.delegate = self
+        for url in urls {
+            guard let fileDirectory = UrlParser.parseForFileDirectory(with: url) else {
+                self.completionHandler?(.failure(NSError.downloadingFailed()))
+                return
+            }
+            urlToDirectoryMap[url] = miniAppPath.appendingPathComponent(fileDirectory)
             queue.addOperation {
-                let task = self.session.downloadTask(with: downloadUrl)
-                task.resume()
+                self.miniAppClient?.download(url: url)
             }
         }
     }
 }
 
-/**
- * URLSessionDownloadDelegate methods for MiniAppDownloader.
- */
-extension MiniAppDownloader: URLSessionDownloadDelegate {
-    func urlSession(_ session: URLSession, downloadTask: URLSessionDownloadTask, didFinishDownloadingTo location: URL) {
-        guard let sourceURL = downloadTask.originalRequest?.url else {
+extension MiniAppDownloader: MiniAppDownloaderProtocol {
+
+    func downloadedFileData(downloadTask: URLSessionDownloadTask, location: URL) {
+        guard let sourceURL = downloadTask.currentRequest?.url else {
             return
         }
-
         guard let destinationPath = urlToDirectoryMap[sourceURL.absoluteString] else {
-            #if DEBUG
-                print("MiniAppSDK: Failed to save files.")
-            #endif
             return
         }
+        miniAppStorage.save(sourcePath: location, destinationPath: destinationPath)
+    }
 
-        try? FileManager.default.createDirectory(atPath: destinationPath.relativePath, withIntermediateDirectories: true)
-        try? FileManager.default.removeItem(at: destinationPath)
-        try? FileManager.default.copyItem(at: location, to: destinationPath)
-
-        urlToDirectoryMap.removeValue(forKey: sourceURL.absoluteString)
+    func downloadCompleted(task: URLSessionTask, error: Error?) {
```

Comment:
This seems to just need a URL passed in?

Reply:
Done